to_parquet failing on pyarrow master when different partitions result in different parquet schema

In pyarrow master, we made a change to `append_row_groups` to require equal schema's when appending (https://issues.apache.org/jira/browse/ARROW-8062, 
https://github.com/apache/arrow/pull/7180). But that apparently causes test failures in the dask test suite:

<details>

```
2020-05-26T06:57:29.5507643Z =================================== FAILURES ===================================
2020-05-26T06:57:29.5509731Z _ test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_fails_by_default __
2020-05-26T06:57:29.5509974Z 
2020-05-26T06:57:29.5512116Z tmpdir = local('/tmp/pytest-of-root/pytest-0/test_to_parquet_pyarrow_w_inco0')
2020-05-26T06:57:29.5512377Z 
2020-05-26T06:57:29.5514934Z     def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_fails_by_default(tmpdir):
2020-05-26T06:57:29.5515252Z         check_pyarrow()
2020-05-26T06:57:29.5515584Z     
2020-05-26T06:57:29.5515752Z         df = pd.DataFrame(
2020-05-26T06:57:29.5515998Z             {"partition_column": [0, 0, 1, 1], "strings": ["a", "b", None, None]}
2020-05-26T06:57:29.5516182Z         )
2020-05-26T06:57:29.5516369Z     
2020-05-26T06:57:29.5516616Z         ddf = dd.from_pandas(df, npartitions=2)
2020-05-26T06:57:29.5516806Z >       ddf.to_parquet(str(tmpdir), engine="pyarrow", partition_on=["partition_column"])
2020-05-26T06:57:29.5516894Z 
2020-05-26T06:57:29.5519255Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/tests/test_parquet.py:965: 
2020-05-26T06:57:29.5519532Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-05-26T06:57:29.5520327Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/core.py:3811: in to_parquet
2020-05-26T06:57:29.5525402Z     return to_parquet(self, path, *args, **kwargs)
2020-05-26T06:57:29.5528566Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:461: in to_parquet
2020-05-26T06:57:29.5529800Z     out = out.compute(**compute_kwargs)
2020-05-26T06:57:29.5530394Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/base.py:166: in compute
2020-05-26T06:57:29.5530739Z     (result,) = compute(self, traverse=False, **kwargs)
2020-05-26T06:57:29.5531158Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/base.py:444: in compute
2020-05-26T06:57:29.5539865Z     results = schedule(dsk, keys, **kwargs)
2020-05-26T06:57:29.5543266Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/threaded.py:84: in get
2020-05-26T06:57:29.5545361Z     **kwargs
2020-05-26T06:57:29.5547293Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:486: in get_async
2020-05-26T06:57:29.5549251Z     raise_exception(exc, tb)
2020-05-26T06:57:29.5550734Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:316: in reraise
2020-05-26T06:57:29.5552431Z     raise exc
2020-05-26T06:57:29.5553073Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:222: in execute_task
2020-05-26T06:57:29.5553366Z     result = _execute_task(task, data)
2020-05-26T06:57:29.5553764Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/core.py:121: in _execute_task
2020-05-26T06:57:29.5553938Z     return func(*(_execute_task(a, cache) for a in args))
2020-05-26T06:57:29.5554268Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/utils.py:30: in apply
2020-05-26T06:57:29.5554462Z     return func(*args, **kwargs)
2020-05-26T06:57:29.5554815Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py:763: in write_metadata
2020-05-26T06:57:29.5554989Z     _meta.append_row_groups(parts[i][0]["meta"])
2020-05-26T06:57:29.5555156Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-05-26T06:57:29.5555299Z 
2020-05-26T06:57:29.5555451Z >   ???
2020-05-26T06:57:29.5555610Z E   RuntimeError: AppendRowGroups requires equal schemas.
2020-05-26T06:57:29.5555722Z 
2020-05-26T06:57:29.5555859Z pyarrow/_parquet.pyx:687: RuntimeError
2020-05-26T06:57:29.5556038Z ___________________ test_timeseries_nulls_in_schema[pyarrow] ___________________
2020-05-26T06:57:29.5556155Z 
2020-05-26T06:57:29.5556469Z tmpdir = local('/tmp/pytest-of-root/pytest-0/test_timeseries_nulls_in_schem0')
2020-05-26T06:57:29.5556765Z engine = 'pyarrow'
2020-05-26T06:57:29.5556881Z 
2020-05-26T06:57:29.5557045Z     def test_timeseries_nulls_in_schema(tmpdir, engine):
2020-05-26T06:57:29.5557218Z         # GH#5608: relative path failing _metadata/_common_metadata detection.
2020-05-26T06:57:29.5557373Z         tmp_path = str(tmpdir.mkdir("files"))
2020-05-26T06:57:29.5557554Z         tmp_path = os.path.join(tmp_path, "../", "files")
2020-05-26T06:57:29.5557698Z     
2020-05-26T06:57:29.5557871Z         ddf2 = (
2020-05-26T06:57:29.5558194Z             dask.datasets.timeseries(start="2000-01-01", end="2000-01-03", freq="1h")
2020-05-26T06:57:29.5558363Z             .reset_index()
2020-05-26T06:57:29.5558526Z             .map_partitions(lambda x: x.loc[:5])
2020-05-26T06:57:29.5558701Z         )
2020-05-26T06:57:29.5558847Z         ddf2 = ddf2.set_index("x").reset_index().persist()
2020-05-26T06:57:29.5559172Z         ddf2.name = ddf2.name.where(ddf2.timestamp == "2000-01-01", None)
2020-05-26T06:57:29.5559356Z     
2020-05-26T06:57:29.5559509Z >       ddf2.to_parquet(tmp_path, engine=engine)
2020-05-26T06:57:29.5559623Z 
2020-05-26T06:57:29.5560119Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/tests/test_parquet.py:2011: 
2020-05-26T06:57:29.5560347Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-05-26T06:57:29.5560754Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/core.py:3811: in to_parquet
2020-05-26T06:57:29.5561517Z     return to_parquet(self, path, *args, **kwargs)
2020-05-26T06:57:29.5562388Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:461: in to_parquet
2020-05-26T06:57:29.5562518Z     out = out.compute(**compute_kwargs)
2020-05-26T06:57:29.5562799Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/base.py:166: in compute
2020-05-26T06:57:29.5562906Z     (result,) = compute(self, traverse=False, **kwargs)
2020-05-26T06:57:29.5563189Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/base.py:444: in compute
2020-05-26T06:57:29.5563309Z     results = schedule(dsk, keys, **kwargs)
2020-05-26T06:57:29.5563699Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/threaded.py:84: in get
2020-05-26T06:57:29.5563815Z     **kwargs
2020-05-26T06:57:29.5564074Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:486: in get_async
2020-05-26T06:57:29.5564193Z     raise_exception(exc, tb)
2020-05-26T06:57:29.5564468Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:316: in reraise
2020-05-26T06:57:29.5564589Z     raise exc
2020-05-26T06:57:29.5564881Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/local.py:222: in execute_task
2020-05-26T06:57:29.5565003Z     result = _execute_task(task, data)
2020-05-26T06:57:29.5565286Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/core.py:121: in _execute_task
2020-05-26T06:57:29.5565410Z     return func(*(_execute_task(a, cache) for a in args))
2020-05-26T06:57:29.5565688Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/utils.py:30: in apply
2020-05-26T06:57:29.5565789Z     return func(*args, **kwargs)
2020-05-26T06:57:29.5566173Z opt/conda/envs/arrow/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py:763: in write_metadata
2020-05-26T06:57:29.5566299Z     _meta.append_row_groups(parts[i][0]["meta"])
2020-05-26T06:57:29.5566410Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-05-26T06:57:29.5566474Z 
2020-05-26T06:57:29.5566570Z >   ???
2020-05-26T06:57:29.5566675Z E   RuntimeError: AppendRowGroups requires equal schemas.
2020-05-26T06:57:29.5566733Z 
2020-05-26T06:57:29.5566831Z pyarrow/_parquet.pyx:687: RuntimeError
```

</details>


I when doing `FileMetadata.append_row_groups`, it *seems* to make sense that we require the schemas to be the same, because afterwards there is only a single schema.

Looking at the failures in dask, it are cases where the different partitions of the dask dataframe result in a different pyarrow/parquet schema (and specifically a partition with all nulls in an object dtype (string) column). 

Do you see a solution for this on the dask side? (in principle there is the `meta` dtypes, but here of course it's the typical problem that "object" dtype is not specific enough ..).

Or should we revert back to being less strict on Arrow's side?

cc @rjzamora @martindurant

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -20,6 +20,7 @@ from .utils import (
 )
 
 preserve_ind_supported = pa.__version__ >= LooseVersion("0.15.0")
+schema_field_supported = pa.__version__ >= LooseVersion("0.15.0")
 
 
 #
@@ -844,12 +845,11 @@ class ArrowEngine(Engine):
         if schema == "infer" or isinstance(schema, dict):
 
             # Start with schema from _meta_nonempty
-            _meta = (
+            _schema = pa.Schema.from_pandas(
                 df._meta_nonempty.set_index(index_cols)
                 if index_cols
                 else df._meta_nonempty
             )
-            _schema = pa.Schema.from_pandas(_meta)
 
             # Use dict to update our inferred schema
             if isinstance(schema, dict):
@@ -862,7 +862,7 @@ class ArrowEngine(Engine):
             # If we have object columns, we need to sample partitions
             # until we find non-null data for each column in `sample`
             sample = [col for col in df.columns if df[col].dtype == "object"]
-            if sample and schema == "infer":
+            if schema_field_supported and sample and schema == "infer":
                 for i in range(df.npartitions):
                     _df = df[sample].get_partition(i)
                     size = len(_df)

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -864,9 +864,10 @@ class ArrowEngine(Engine):
             # until we find non-null data for each column in `sample`
             sample = [col for col in df.columns if df[col].dtype == "object"]
             if schema_field_supported and sample and schema == "infer":
+                delayed_schema_from_pandas = delayed(pa.Schema.from_pandas)
                 for i in range(df.npartitions):
                     # Keep data on worker
-                    _s = delayed(lambda x: pa.Schema.from_pandas(x))(
+                    _s = delayed_schema_from_pandas(
                         df[sample].to_delayed()[i]
                     ).compute()
                     for name, typ in zip(_s.names, _s.types):

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -821,8 +821,36 @@ class ArrowEngine(Engine):
         partition_on=None,
         ignore_divisions=False,
         division_info=None,
+        schema="sample",
+        index_cols=None,
         **kwargs,
     ):
+        # Infer schema if "sample" or "meta"...
+        # Note that "meta" is faster, but may fail to produce
+        # the correct schema for "object" types.
+        if schema == "sample":
+            # Use first partition with data to infer schema
+            for i in range(df.npartitions):
+                _df = df.get_partition(i)
+                size = len(_df)
+                if size:
+                    _df = _df.head(min(size, 5))
+                    break
+            schema = None
+            if size:
+                # We found data, reset schema with pyarrow
+                schema = pa.Schema.from_pandas(
+                    _df.set_index(index_cols) if index_cols else _df
+                )
+                del _df
+        elif schema == "meta":
+            # Use _meta_nonempty to infer schema
+            schema = pa.Schema.from_pandas(
+                df._meta_nonempty.set_index(index_cols)
+                if index_cols
+                else df._meta_nonempty
+            )
+
         dataset = fmd = None
         i_offset = 0
         if append and division_info is None:
@@ -906,7 +934,7 @@ class ArrowEngine(Engine):
                         "Previous: {} | New: {}".format(old_end, divisions[0])
                     )
 
-        return fmd, i_offset
+        return fmd, schema, i_offset
 
     @staticmethod
     def write_partition(

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -840,7 +840,8 @@ class ArrowEngine(Engine):
         **kwargs,
     ):
         # Infer schema if "infer"
-        if schema == "infer":
+        # (also start with inferred schema if user passes a dict)
+        if schema == "infer" or isinstance(schema, dict):
 
             # Start with schema from _meta_nonempty
             _meta = (
@@ -848,12 +849,20 @@ class ArrowEngine(Engine):
                 if index_cols
                 else df._meta_nonempty
             )
-            schema = pa.Schema.from_pandas(_meta)
-            sample = [col for col in df.columns if df[col].dtype == "object"]
+            _schema = pa.Schema.from_pandas(_meta)
+
+            # Use dict to update our inferred schema
+            if isinstance(schema, dict):
+                schema = pa.schema(schema)
+                for name in schema.names:
+                    i = _schema.get_field_index(name)
+                    j = schema.get_field_index(name)
+                    _schema.set(i, schema.field(j))
 
             # If we have object columns, we need to sample partitions
             # until we find non-null data for each column in `sample`
-            if sample:
+            sample = [col for col in df.columns if df[col].dtype == "object"]
+            if sample and schema == "infer":
                 for i in range(df.npartitions):
                     _df = df[sample].get_partition(i)
                     size = len(_df)
@@ -861,12 +870,15 @@ class ArrowEngine(Engine):
                         _s = pa.Schema.from_pandas(_df.compute())
                         for name, typ in zip(_s.names, _s.types):
                             if typ != "null":
-                                i = schema.get_field_index(name)
+                                i = _schema.get_field_index(name)
                                 j = _s.get_field_index(name)
-                                schema.set(i, _s.field(j))
+                                _schema.set(i, _s.field(j))
                                 sample.remove(name)
                     if not sample:
                         break
+
+            # Final (inferred) schema
+            schema = _schema
 
         dataset = fmd = None
         i_offset = 0

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -872,7 +872,7 @@ class ArrowEngine(Engine):
                             if typ != "null":
                                 i = _schema.get_field_index(name)
                                 j = _s.get_field_index(name)
-                                _schema.set(i, _s.field(j))
+                                _schema = _schema.set(i, _s.field(j))
                                 sample.remove(name)
                     if not sample:
                         break

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -11,6 +11,7 @@ from ....utils import getargspec
 from ..utils import _get_pyarrow_dtypes, _meta_from_dtypes
 from ...utils import clear_known_categories
 from ....core import flatten
+from dask import delayed
 
 from .utils import (
     _parse_pandas_metadata,
@@ -864,16 +865,16 @@ class ArrowEngine(Engine):
             sample = [col for col in df.columns if df[col].dtype == "object"]
             if schema_field_supported and sample and schema == "infer":
                 for i in range(df.npartitions):
-                    _df = df[sample].get_partition(i)
-                    size = len(_df)
-                    if size:
-                        _s = pa.Schema.from_pandas(_df.compute())
-                        for name, typ in zip(_s.names, _s.types):
-                            if typ != "null":
-                                i = _schema.get_field_index(name)
-                                j = _s.get_field_index(name)
-                                _schema = _schema.set(i, _s.field(j))
-                                sample.remove(name)
+                    # Keep data on worker
+                    _s = delayed(lambda x: pa.Schema.from_pandas(x))(
+                        df[sample].to_delayed()[i]
+                    ).compute()
+                    for name, typ in zip(_s.names, _s.types):
+                        if typ != "null":
+                            i = _schema.get_field_index(name)
+                            j = _s.get_field_index(name)
+                            _schema = _schema.set(i, _s.field(j))
+                            sample.remove(name)
                     if not sample:
                         break
 

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -27,6 +27,20 @@ preserve_ind_supported = pa.__version__ >= LooseVersion("0.15.0")
 #
 
 
+def _append_row_groups(metadata, md):
+    try:
+        metadata.append_row_groups(md)
+    except RuntimeError as err:
+        if "requires equal schemas" in str(err):
+            raise RuntimeError(
+                "Schemas are inconsistent, try using "
+                '`to_parquet(..., schema="infer")`, or pass an explicit '
+                "pyarrow schema."
+            )
+        else:
+            raise err
+
+
 def _write_partitioned(
     table, root_path, filename, partition_cols, fs, index_cols=(), **kwargs
 ):
@@ -239,7 +253,7 @@ def _gather_metadata(
             elif fn:
                 md.set_file_path(fn)
             if metadata:
-                metadata.append_row_groups(md)
+                _append_row_groups(metadata, md)
             else:
                 metadata = md
         return (
@@ -821,35 +835,38 @@ class ArrowEngine(Engine):
         partition_on=None,
         ignore_divisions=False,
         division_info=None,
-        schema="sample",
+        schema=None,
         index_cols=None,
         **kwargs,
     ):
-        # Infer schema if "sample" or "meta"...
-        # Note that "meta" is faster, but may fail to produce
-        # the correct schema for "object" types.
-        if schema == "sample":
-            # Use first partition with data to infer schema
-            for i in range(df.npartitions):
-                _df = df.get_partition(i)
-                size = len(_df)
-                if size:
-                    _df = _df.head(min(size, 5))
-                    break
-            schema = None
-            if size:
-                # We found data, reset schema with pyarrow
-                schema = pa.Schema.from_pandas(
-                    _df.set_index(index_cols) if index_cols else _df
-                )
-                del _df
-        elif schema == "meta":
-            # Use _meta_nonempty to infer schema
-            schema = pa.Schema.from_pandas(
+        # Infer schema if "infer"
+        if schema == "infer":
+
+            # Start with schema from _meta_nonempty
+            _meta = (
                 df._meta_nonempty.set_index(index_cols)
                 if index_cols
                 else df._meta_nonempty
             )
+            schema = pa.Schema.from_pandas(_meta)
+            sample = [col for col in df.columns if df[col].dtype == "object"]
+
+            # If we have object columns, we need to sample partitions
+            # until we find non-null data for each column in `sample`
+            if sample:
+                for i in range(df.npartitions):
+                    _df = df[sample].get_partition(i)
+                    size = len(_df)
+                    if size:
+                        _s = pa.Schema.from_pandas(_df.compute())
+                        for name, typ in zip(_s.names, _s.types):
+                            if typ != "null":
+                                i = schema.get_field_index(name)
+                                j = _s.get_field_index(name)
+                                schema.set(i, _s.field(j))
+                                sample.remove(name)
+                    if not sample:
+                        break
 
         dataset = fmd = None
         i_offset = 0
@@ -965,7 +982,7 @@ class ArrowEngine(Engine):
             if md_list:
                 _meta = md_list[0]
                 for i in range(1, len(md_list)):
-                    _meta.append_row_groups(md_list[i])
+                    _append_row_groups(_meta, md_list[i])
         else:
             md_list = []
             with fs.open(fs.sep.join([path, filename]), "wb") as fil:
@@ -1005,6 +1022,6 @@ class ArrowEngine(Engine):
                 _meta = parts[0][0]["meta"]
                 i_start = 1
             for i in range(i_start, len(parts)):
-                _meta.append_row_groups(parts[i][0]["meta"])
+                _append_row_groups(_meta, parts[i][0]["meta"])
             with fs.open(metadata_path, "wb") as fil:
                 _meta.write_metadata_file(fil)

--- a/dask/dataframe/io/parquet/arrow.py
+++ b/dask/dataframe/io/parquet/arrow.py
@@ -857,7 +857,7 @@ class ArrowEngine(Engine):
                 for name in schema.names:
                     i = _schema.get_field_index(name)
                     j = schema.get_field_index(name)
-                    _schema.set(i, schema.field(j))
+                    _schema = _schema.set(i, schema.field(j))
 
             # If we have object columns, we need to sample partitions
             # until we find non-null data for each column in `sample`

--- a/dask/dataframe/io/parquet/core.py
+++ b/dask/dataframe/io/parquet/core.py
@@ -98,7 +98,7 @@ def read_parquet(
     gather_statistics=None,
     split_row_groups=None,
     chunksize=None,
-    **kwargs
+    **kwargs,
 ):
     """
     Read a Parquet file into a Dask DataFrame
@@ -236,7 +236,7 @@ def read_parquet(
         gather_statistics=gather_statistics,
         filters=filters,
         split_row_groups=split_row_groups,
-        **kwargs
+        **kwargs,
     )
 
     # Parse dataset statistics from metadata (if available)
@@ -301,7 +301,8 @@ def to_parquet(
     write_metadata_file=True,
     compute=True,
     compute_kwargs=None,
-    **kwargs
+    schema="sample",
+    **kwargs,
 ):
     """Store Dask.dataframe to Parquet files
 
@@ -345,6 +346,13 @@ def to_parquet(
         then a ``dask.delayed`` object is returned for future computation.
     compute_kwargs : dict, optional
         Options to be passed in to the compute method
+    schema : Schema object, or {"sample", "meta", None}, optional
+        Global schema to use for the output dataset. If "sample" (default),
+        the `head` of the first non-empty partition will be used to infer
+        the global schema. If "meta", `_meta_nonempty` will be used to infer
+        the global schema. If None, we let the backend infer the schema
+        for each distinct output partition. Note that this argument is
+        currently ignored by the "fastparquet" engine.
     **kwargs :
         Extra options to be passed on to the specific backend.
 
@@ -404,6 +412,14 @@ def to_parquet(
                 "will be set to the index (and renamed to None)."
             )
 
+    # There are some "resrved" names that may be used as the default column
+    # name after resetting the index. However, we don't want to treat it as
+    # a "special" name if the string is already used as a "real" column name.
+    reserved_names = []
+    for name in ["index", "level_0"]:
+        if name not in df.columns:
+            reserved_names.append(name)
+
     # If write_index==True (default), reset the index and record the
     # name of the original index in `index_cols` (we will set the name
     # to the NONE_LABEL constant if it is originally `None`).
@@ -418,7 +434,9 @@ def to_parquet(
         none_index = list(df._meta.index.names) == [None]
         df = df.reset_index()
         if none_index:
-            df.columns = [c if c != "index" else NONE_LABEL for c in df.columns]
+            df.columns = [
+                c if c not in reserved_names else NONE_LABEL for c in df.columns
+            ]
         index_cols = [c for c in set(df.columns).difference(real_cols)]
     else:
         # Not writing index - might as well drop it
@@ -439,7 +457,7 @@ def to_parquet(
 
     # Engine-specific initialization steps to write the dataset.
     # Possibly create parquet metadata, and load existing stuff if appending
-    meta, i_offset = engine.initialize_write(
+    meta, schema, i_offset = engine.initialize_write(
         df,
         fs,
         path,
@@ -448,7 +466,8 @@ def to_parquet(
         partition_on=partition_on,
         division_info=division_info,
         index_cols=index_cols,
-        **kwargs_pass
+        schema=schema,
+        **kwargs_pass,
     )
 
     # Use i_offset and df.npartitions to define file-name list
@@ -467,7 +486,8 @@ def to_parquet(
             fmd=meta,
             compression=compression,
             index_cols=index_cols,
-            **kwargs_pass
+            schema=schema,
+            **kwargs_pass,
         )
         for d, filename in zip(df.to_delayed(), filenames)
     ]

--- a/dask/dataframe/io/parquet/core.py
+++ b/dask/dataframe/io/parquet/core.py
@@ -301,7 +301,7 @@ def to_parquet(
     write_metadata_file=True,
     compute=True,
     compute_kwargs=None,
-    schema="sample",
+    schema=None,
     **kwargs,
 ):
     """Store Dask.dataframe to Parquet files
@@ -346,11 +346,10 @@ def to_parquet(
         then a ``dask.delayed`` object is returned for future computation.
     compute_kwargs : dict, optional
         Options to be passed in to the compute method
-    schema : Schema object, or {"sample", "meta", None}, optional
-        Global schema to use for the output dataset. If "sample" (default),
-        the `head` of the first non-empty partition will be used to infer
-        the global schema. If "meta", `_meta_nonempty` will be used to infer
-        the global schema. If None, we let the backend infer the schema
+    schema : Schema object, or {"infer", None}, optional
+        Global schema to use for the output dataset. If "infer", the `head`
+        of the first non-empty partition will be used to infer the global
+        schema. If None (default), we let the backend infer the schema
         for each distinct output partition. Note that this argument is
         currently ignored by the "fastparquet" engine.
     **kwargs :

--- a/dask/dataframe/io/parquet/core.py
+++ b/dask/dataframe/io/parquet/core.py
@@ -346,12 +346,15 @@ def to_parquet(
         then a ``dask.delayed`` object is returned for future computation.
     compute_kwargs : dict, optional
         Options to be passed in to the compute method
-    schema : Schema object, or {"infer", None}, optional
-        Global schema to use for the output dataset. If "infer", the `head`
-        of the first non-empty partition will be used to infer the global
-        schema. If None (default), we let the backend infer the schema
-        for each distinct output partition. Note that this argument is
-        currently ignored by the "fastparquet" engine.
+    schema : Schema object, dict, or {"infer", None}, optional
+        Global schema to use for the output dataset. Alternatively, a `dict`
+        of pyarrow types can be specified (e.g. `schema={"id": pa.string()}`).
+        For this case, fields excluded from the dictionary will be inferred
+        from `_meta_nonempty`.  If "infer", the first non-empty and non-null
+        partition will be used to infer the type for "object" columns. If
+        None (default), we let the backend infer the schema for each distinct
+        output partition. Note that this argument is currently ignored by
+        the "fastparquet" engine.
     **kwargs :
         Extra options to be passed on to the specific backend.
 

--- a/dask/dataframe/io/parquet/core.py
+++ b/dask/dataframe/io/parquet/core.py
@@ -353,8 +353,9 @@ def to_parquet(
         from `_meta_nonempty`.  If "infer", the first non-empty and non-null
         partition will be used to infer the type for "object" columns. If
         None (default), we let the backend infer the schema for each distinct
-        output partition. Note that this argument is currently ignored by
-        the "fastparquet" engine.
+        output partition. If the partitions produce inconsistent schemas,
+        pyarrow will throw an error when writing the shared _metadata file.
+        Note that this argument is ignored by the "fastparquet" engine.
     **kwargs :
         Extra options to be passed on to the specific backend.
 

--- a/dask/dataframe/io/parquet/fastparquet.py
+++ b/dask/dataframe/io/parquet/fastparquet.py
@@ -510,6 +510,7 @@ class FastParquetEngine(Engine):
         partition_on=None,
         ignore_divisions=False,
         division_info=None,
+        schema=None,
         **kwargs
     ):
         if append and division_info is None:
@@ -579,7 +580,8 @@ class FastParquetEngine(Engine):
             )
             i_offset = 0
 
-        return (fmd, i_offset)
+        schema = None  # ArrowEngine compatibility
+        return (fmd, schema, i_offset)
 
     @classmethod
     def write_partition(

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -1074,8 +1074,7 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_succeeds_w_manual
 
 @pytest.mark.parametrize("index", [False, True])
 @pytest.mark.parametrize(
-    "schema",
-    ["infer", {"index": pa.string(), "date": pa.string(), "amount": pa.int64()}],
+    "schema", ["infer", {"index": pa.string(), "amount": pa.int64()}]
 )
 def test_pyarrow_schema_inference(tmpdir, index, engine, schema):
 
@@ -1107,7 +1106,9 @@ def test_pyarrow_schema_inference(tmpdir, index, engine, schema):
     else:
         df = dd.from_pandas(df, npartitions=2)
 
-    df.to_parquet(tmpdir, engine="pyarrow", schema=schema)
+    df.to_parquet(tmpdir, engine="pyarrow", schema=schema, compute=False).compute(
+        scheduler="synchronous"
+    )
     df_out = dd.read_parquet(tmpdir, engine=engine)
 
     if index and engine == "fastparquet":

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -1074,7 +1074,11 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_succeeds_w_manual
 
 @write_read_engines()
 @pytest.mark.parametrize("index", [False, True])
-def test_schema_inference(tmpdir, index, write_engine, read_engine):
+@pytest.mark.parametrize(
+    "schema",
+    ["infer", {"index": pa.string(), "date": pa.string(), "amount": pa.int64()}],
+)
+def test_schema_inference(tmpdir, index, write_engine, read_engine, schema):
 
     tmpdir = str(tmpdir)
     df = pd.DataFrame(
@@ -1100,7 +1104,7 @@ def test_schema_inference(tmpdir, index, write_engine, read_engine):
     else:
         df = dd.from_pandas(df, npartitions=2)
 
-    df.to_parquet(tmpdir, engine=write_engine, schema="infer")
+    df.to_parquet(tmpdir, engine=write_engine, schema=schema)
     df_out = dd.read_parquet(tmpdir, engine=read_engine)
 
     if index and read_engine == "fastparquet":

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -1074,8 +1074,7 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_succeeds_w_manual
 
 @write_read_engines()
 @pytest.mark.parametrize("index", [False, True])
-@pytest.mark.parametrize("schema", ["meta", "sample"])
-def test_schema_inference(tmpdir, index, schema, write_engine, read_engine):
+def test_schema_inference(tmpdir, index, write_engine, read_engine):
 
     tmpdir = str(tmpdir)
     df = pd.DataFrame(
@@ -1101,7 +1100,7 @@ def test_schema_inference(tmpdir, index, schema, write_engine, read_engine):
     else:
         df = dd.from_pandas(df, npartitions=2)
 
-    df.to_parquet(tmpdir, engine=write_engine, schema=schema)
+    df.to_parquet(tmpdir, engine=write_engine, schema="infer")
     df_out = dd.read_parquet(tmpdir, engine=read_engine)
 
     if index and read_engine == "fastparquet":
@@ -2072,7 +2071,7 @@ def test_read_dir_nometa(tmpdir, write_engine, read_engine, statistics, remove_c
     assert_eq(ddf, ddf2, check_divisions=False)
 
 
-@pytest.mark.parametrize("schema", ["meta", None])
+@pytest.mark.parametrize("schema", ["infer", None])
 def test_timeseries_nulls_in_schema(tmpdir, engine, schema):
     # GH#5608: relative path failing _metadata/_common_metadata detection.
     tmp_path = str(tmpdir.mkdir("files"))

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -960,12 +960,15 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_fails_by_default(
     )
 
     ddf = dd.from_pandas(df, npartitions=2)
-    # Note: `append_row_groups` will fail with pyarrow>17.1 for _metadata write
+    # In order to allow pyarrow to write an inconsistent schema,
+    # we need to avoid writing the _metadata file (will fail >0.17.1)
+    # and need to avoid schema inference (i.e. use `schema=None`)
     ddf.to_parquet(
         str(tmpdir),
         engine="pyarrow",
         partition_on=["partition_column"],
         write_metadata_file=False,
+        schema=None,
     )
 
     # Test that read fails because of default behavior when schema not provided
@@ -1067,6 +1070,46 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_succeeds_w_manual
 
     # Check partition column
     assert np.array_equal(ddf_after_write.partition_column, df.partition_column)
+
+
+@write_read_engines()
+@pytest.mark.parametrize("index", [False, True])
+@pytest.mark.parametrize("schema", ["meta", "sample"])
+def test_schema_inference(tmpdir, index, schema, write_engine, read_engine):
+
+    tmpdir = str(tmpdir)
+    df = pd.DataFrame(
+        {
+            "index": ["1", "2", "3", "2", "3", "1", "4"],
+            "date": pd.to_datetime(
+                [
+                    "2017-01-01",
+                    "2017-01-01",
+                    "2017-01-01",
+                    "2017-01-02",
+                    "2017-01-02",
+                    "2017-01-06",
+                    "2017-01-09",
+                ]
+            ),
+            "amount": [100, 200, 300, 400, 500, 600, 700],
+        },
+        index=range(7, 14),
+    )
+    if index:
+        df = dd.from_pandas(df, npartitions=2).set_index("index")
+    else:
+        df = dd.from_pandas(df, npartitions=2)
+
+    df.to_parquet(tmpdir, engine=write_engine, schema=schema)
+    df_out = dd.read_parquet(tmpdir, engine=read_engine)
+
+    if index and read_engine == "fastparquet":
+        # Fastparquet not handling divisions for
+        # pyarrow-written dataset with string index
+        assert_eq(df, df_out, check_divisions=False)
+    else:
+        assert_eq(df, df_out)
 
 
 def test_partition_on(tmpdir, engine):
@@ -2029,7 +2072,8 @@ def test_read_dir_nometa(tmpdir, write_engine, read_engine, statistics, remove_c
     assert_eq(ddf, ddf2, check_divisions=False)
 
 
-def test_timeseries_nulls_in_schema(tmpdir, engine):
+@pytest.mark.parametrize("schema", ["meta", None])
+def test_timeseries_nulls_in_schema(tmpdir, engine, schema):
     # GH#5608: relative path failing _metadata/_common_metadata detection.
     tmp_path = str(tmpdir.mkdir("files"))
     tmp_path = os.path.join(tmp_path, "../", "files")
@@ -2043,7 +2087,7 @@ def test_timeseries_nulls_in_schema(tmpdir, engine):
     ddf2.name = ddf2.name.where(ddf2.timestamp == "2000-01-01", None)
 
     # Note: `append_row_groups` will fail with pyarrow>0.17.1 for _metadata write
-    ddf2.to_parquet(tmp_path, engine=engine, write_metadata_file=False)
+    ddf2.to_parquet(tmp_path, engine=engine, write_metadata_file=False, schema=schema)
     ddf_read = dd.read_parquet(
         tmp_path, engine=engine, dataset={"validate_schema": False}
     )
@@ -2051,8 +2095,9 @@ def test_timeseries_nulls_in_schema(tmpdir, engine):
     assert_eq(ddf_read, ddf2, check_divisions=False, check_index=False)
 
     # Can force schema validation on each partition in pyarrow
-    if engine == "pyarrow":
-        # The schema mismatch should raise an error
+    if engine == "pyarrow" and schema is None:
+        # The schema mismatch should raise an error if the
+        # dataset was written with `schema=None` (no inference)
         with pytest.raises(ValueError):
             ddf_read = dd.read_parquet(
                 tmp_path, dataset={"validate_schema": True}, engine=engine

--- a/dask/dataframe/io/tests/test_parquet.py
+++ b/dask/dataframe/io/tests/test_parquet.py
@@ -1072,13 +1072,16 @@ def test_to_parquet_pyarrow_w_inconsistent_schema_by_partition_succeeds_w_manual
     assert np.array_equal(ddf_after_write.partition_column, df.partition_column)
 
 
-@write_read_engines()
 @pytest.mark.parametrize("index", [False, True])
 @pytest.mark.parametrize(
     "schema",
     ["infer", {"index": pa.string(), "date": pa.string(), "amount": pa.int64()}],
 )
-def test_schema_inference(tmpdir, index, write_engine, read_engine, schema):
+def test_pyarrow_schema_inference(tmpdir, index, engine, schema):
+
+    check_pyarrow()
+    if pa.__version__ < LooseVersion("0.15.0"):
+        pytest.skip("PyArrow>=0.15 Required.")
 
     tmpdir = str(tmpdir)
     df = pd.DataFrame(
@@ -1104,10 +1107,10 @@ def test_schema_inference(tmpdir, index, write_engine, read_engine, schema):
     else:
         df = dd.from_pandas(df, npartitions=2)
 
-    df.to_parquet(tmpdir, engine=write_engine, schema=schema)
-    df_out = dd.read_parquet(tmpdir, engine=read_engine)
+    df.to_parquet(tmpdir, engine="pyarrow", schema=schema)
+    df_out = dd.read_parquet(tmpdir, engine=engine)
 
-    if index and read_engine == "fastparquet":
+    if index and engine == "fastparquet":
         # Fastparquet not handling divisions for
         # pyarrow-written dataset with string index
         assert_eq(df, df_out, check_divisions=False)
@@ -2077,6 +2080,14 @@ def test_read_dir_nometa(tmpdir, write_engine, read_engine, statistics, remove_c
 
 @pytest.mark.parametrize("schema", ["infer", None])
 def test_timeseries_nulls_in_schema(tmpdir, engine, schema):
+
+    if (
+        schema == "infer"
+        and engine == "pyarrow"
+        and pa.__version__ < LooseVersion("0.15.0")
+    ):
+        pytest.skip("PyArrow>=0.15 Required.")
+
     # GH#5608: relative path failing _metadata/_common_metadata detection.
     tmp_path = str(tmpdir.mkdir("files"))
     tmp_path = os.path.join(tmp_path, "../", "files")

